src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,240 @@
+    }
+
+    @Override
+    public int[] getRequiredJavadocTokens() {
+        return getAcceptableJavadocTokens();
+    }
+
+    @Override
+    public void beginJavadocTree(DetailNode rootAst) {
+        // this method processes and sets information of starting javadoc tag.
+        fileLines = getLines();
+        final String startLine = fileLines[rootAst.getLineNumber() - 1];
+        javadocStartLineNumber = rootAst.getLineNumber();
+        expectedColumnNumber = CommonUtil.lengthExpandedTabs(
+            startLine, rootAst.getColumnNumber() - 1, getTabWidth());
+    }
+
+    @Override
+    public void visitJavadocToken(DetailNode ast) {
+        // this method checks the alignment of leading asterisks.
+        final boolean isJavadocStartingLine = ast.getLineNumber() == javadocStartLineNumber;
+        if (!isJavadocStartingLine) {
+            final OptionalInt asteriskColumnNumber =
+                                        getAsteriskColumnNumberWithTabsExpanded(ast.getText());
+
+            asteriskColumnNumber.ifPresent(integer -> {
+                checkAsteriskAlignment(ast.getLineNumber(),
+                        expectedColumnNumber,
+                        integer);
+            });
+        }
+    }
+
+    @Override
+    public void finishJavadocTree(DetailNode rootAst) {
+        // this method checks the alignment of closing javadoc tag.
+        final DetailAST javadocEndToken = getBlockCommentAst().getLastChild();
+        final String lastLine = fileLines[javadocEndToken.getLineNo() - 1];
+        final OptionalInt asteriskColumnNumber =
+                    getAsteriskColumnNumberWithTabsExpanded(lastLine);
+
+        asteriskColumnNumber.ifPresent(integer -> {
+            checkAsteriskAlignment(javadocEndToken.getLineNo(),
+                    expectedColumnNumber,
+                    integer);
+        });
+    }
+
+    /**
+     * Processes and returns an Optional containing
+     * the column number of leading asterisk with tabs expanded.
+     * Also sets 'singleSpaceAsteriskColNumber' if the leading asterisk is present.
+     *
+     * @param line single line from block comment text
+     * @return column number of leading asterisk with tabs expanded
+     */
+    private OptionalInt getAsteriskColumnNumberWithTabsExpanded(String line) {
+        OptionalInt columnNumber = getAsteriskColumnNumber(line);
+
+        if (columnNumber.isPresent()) {
+            singleSpaceAsteriskColNumber = columnNumber.orElseThrow();
+            columnNumber = OptionalInt.of(CommonUtil.lengthExpandedTabs(
+                        line, columnNumber.orElseThrow(), getTabWidth()));
+        }
+
+        return columnNumber;
+    }
+
+    /**
+     * Processes and returns an Optional containing
+     * the column number of leading asterisk without tabs expanded.
+     *
+     * @param line javadoc comment line
+     * @return asterisk's column number
+     */
+    private static OptionalInt getAsteriskColumnNumber(String line) {
```

Comment:
This should be the only place that we return an Optional from.

Reply:
Done